iothub_client/devdoc/iothubclient_c_library.md

```diff
@@ -475,29 +475,54 @@ By default messages never expire. The meaning of the messageTimeout value is the
     - Any other number - consider that number as the timeout.
 - "x509certificate" - feeds a x509 certificate in PEM format to IoTHubClient to be used for 
-authentication. value is a pointer to a null terminated string that contains the certificate. Example:
+authentication. value is a pointer to a null terminated string that contains the certificate. Important: certain TLS stacks are sensitive to line terminators. Example:
 ```c
 const char* value =
-"-----BEGIN CERTIFICATE-----"
-"MIICpDCCAYwCCQCfIjBnPxs5TzANBgkqhkiG9w0BAQsFADAUMRIwEAYDVQQDDAls"
-"b2NhbGhvc3QwHhcNMTYwNjIyMjM0MzI3WhcNMTYwNjIzMjM0MzI3WjAUMRIwEAYD"
+"-----BEGIN CERTIFICATE-----\n"
+"MIICpDCCAYwCCQCfIjBnPxs5TzANBgkqhkiG9w0BAQsFADAUMRIwEAYDVQQDDAls\n"
+"b2NhbGhvc3QwHhcNMTYwNjIyMjM0MzI3WhcNMTYwNjIzMjM0MzI3WjAUMRIwEAYD\n"
 ...
-"+s88wBF907s1dcY45vsG0ldE3f7Y6anGF60nUwYao/fN/eb5FT5EHANVMmnK8zZ2"
-"tjWUt5TFnAveFoQWIoIbtzlTbOxUFwMrQFzFXOrZoDJmHNWc2u6FmVAkowoOSHiE"
-"dkyVdoGPCXc="
-"-----END CERTIFICATE-----";
+"+s88wBF907s1dcY45vsG0ldE3f7Y6anGF60nUwYao/fN/eb5FT5EHANVMmnK8zZ2\n"
+"tjWUt5TFnAveFoQWIoIbtzlTbOxUFwMrQFzFXOrZoDJmHNWc2u6FmVAkowoOSHiE\n"
+"dkyVdoGPCXc=\n"
+"-----END CERTIFICATE-----\n";
 ```
-- "x509privatekey" - feed a x509 private key in PEM format to IoTHubClient to be used for authentication. value is a pointer to a null terminated string that contains the key. Example:
+- "x509privatekey" - feed a x509 private key in PEM format to IoTHubClient to be used for authentication. value is a pointer to a null terminated string that contains the key. Important: certain TLS stacks are sensitive to line terminators. Example:
 ```c
 const char* privateKey = 
-"-----BEGIN RSA PRIVATE KEY-----"            
-"MIIEpQIBAAKCAQEA0zKK+Uu5I0nXq2V6+2gbdCsBXZ6j1uAgU/clsCohEAek1T8v"
-"qj2tR9Mz9iy9RtXPMHwzcQ7aXDaz7RbHdw7tYXqSw8iq0Mxq2s3p4mo6gd5vEOiN"
+"-----BEGIN RSA PRIVATE KEY-----\n"            
+"MIIEpQIBAAKCAQEA0zKK+Uu5I0nXq2V6+2gbdCsBXZ6j1uAgU/clsCohEAek1T8v\n"
+"qj2tR9Mz9iy9RtXPMHwzcQ7aXDaz7RbHdw7tYXqSw8iq0Mxq2s3p4mo6gd5vEOiN\n"
 ...
-"EyePNmkCgYEAng+12qvs0de7OhkTjX9FLxluLWxfN2vbtQCWXslLCG+Es/ZzGlNF"
-"SaqVID4EAUgUqFDw0UO6SKLT+HyFjOr5qdHkfAmRzwE/0RBN69g2qLDN3Km1Px/k"
-"xyJyxc700uV1eKiCdRLRuCbUeecOSZreh8YRIQQXoG8uotO5IttdVRc="        
-"-----END RSA PRIVATE KEY-----";
+"EyePNmkCgYEAng+12qvs0de7OhkTjX9FLxluLWxfN2vbtQCWXslLCG+Es/ZzGlNF\n"
+"SaqVID4EAUgUqFDw0UO6SKLT+HyFjOr5qdHkfAmRzwE/0RBN69g2qLDN3Km1Px/k\n"
+"xyJyxc700uV1eKiCdRLRuCbUeecOSZreh8YRIQQXoG8uotO5IttdVRc=\n"        
+"-----END RSA PRIVATE KEY-----\n";
+```
+
+- "OPENSSLOPT_ENGINE" - only available when OpenSSL is used. It specifies the [OpenSSL built-in engine](https://www.openssl.org/docs/man1.1.1/man3/ENGINE_load_builtin_engines.html) to be loaded. This option changes the meaning of the `x509certificate` and `x509privatekey` options to be used as the certificate identifier and key identifier respectively. value is a null terminated string that contains the engine name.
```

Comment:
>`x509certificate` [](start = 234, length = 17)

Is this a semantic change or is it one that has implications down the road? 

From my understanding of the EIS flow we are requried to have three things to do fully authenticate a connection:
1. Cert-Id
2. KeyHandle
3. The certificate provided via EIS in pem format.


Would this not mean that we are unable to set the pem string or am I way off kilter?

Reply:
Good point: to avoid any confusion, I've added the two new options `x509CertificateType` and `x509PrivatekeyType`.
By default the engine will be loaded and all its methods enabled. The app needs to specify if the certificate / keys must be loaded from the engine (presenting a key Id) or they are PEM strings (default).